### PR TITLE
[AMLS] FFN and LeNet

### PR DESCRIPTION
**Pull request by aleksapand**
@Baunsgaard paramserv is called in file ffTrain on 234.

### CODE REVIEW

**Comment by Baunsgaard** [comment]
> @Baunsgaard paramserv is called in file ffTrain on 234.

The bug should be fixed now, since it was related to not being able to source from within our builtin functions.

Is this correct?

also the git tests fail in building because some license is missing. use 'mvn package -P rat' and look in the target/rat.txt file to see where it is missing.

**Comment by aleksapand** [comment]
@Baunsgaard Hi Sebastian, thank you for the input. Unfortunately, it still fails.

**Comment by Baunsgaard** [comment]
> @Baunsgaard Hi Sebastian, thank you for the input. Unfortunately, it still fails.

okay will take a look tomorrow.

**Comment by Baunsgaard** [comment]
Okay i have been looking at this, and there is a bug.

The bug for the paramserv cannot be called inside a builtin function since it in turn try parse functions for update and gradients.

If we change the line initializing the parameterserver to:

  model_trained = paramserv(model=model, features=X, labels=y,
    upd="./scripts/builtin/ffTrain.dml::gradients", agg="./scripts/builtin/ffTrain.dml::aggregation", 
    epochs=epochs, hyperparams=params, mode=mode, utype=utype, freq=freq, batchsize=batch_size, k=workers, 
    checkpointing="NONE")

Then it is syntactically correct, but in turn our passer encounter a bug, where it cannot build the ffTrain builtin.

```
[ERROR] Failures: 
[ERROR]   BuiltinFFNeuralNetworkTest.testClassificationFit:91->AutomatedTestBase.runTest:1333->AutomatedTestBase.runTest:1337->AutomatedTestBase.runTest:1341->AutomatedTestBase.runTest:1446 
failed to run script: ./src/test/scripts/functions/builtin/ffNeuralNetwork.dml
Standard Out:

StackTrace:
LEVEL : 0
Exception : class org.apache.sysds.api.DMLException
Message   : failed add function to graph fcall _main
0  >  org.apache.sysds.hops.ipa.FunctionCallGraph.addFunctionOpToGraph(FunctionCallGraph.java:507)
0  >  org.apache.sysds.hops.ipa.FunctionCallGraph.rConstructFunctionCallGraph(FunctionCallGraph.java:424)
0  >  org.apache.sysds.hops.ipa.FunctionCallGraph.constructFunctionCallGraph(FunctionCallGraph.java:360)
0  >  org.apache.sysds.hops.ipa.FunctionCallGraph.<init>(FunctionCallGraph.java:94)
0  >  org.apache.sysds.hops.ipa.InterProceduralAnalysis.<init>(InterProceduralAnalysis.java:133)
0  >  org.apache.sysds.parser.DMLTranslator.rewriteHopsDAG(DMLTranslator.java:285)
0  >  org.apache.sysds.api.DMLScript.execute(DMLScript.java:400)
0  >  org.apache.sysds.api.DMLScript.executeScript(DMLScript.java:269)
0  >  org.apache.sysds.test.AutomatedTestBase.main(AutomatedTestBase.java:1464)
0  >  org.apache.sysds.test.AutomatedTestBase.runTest(AutomatedTestBase.java:1420)
0  >  org.apache.sysds.test.AutomatedTestBase.runTest(AutomatedTestBase.java:1341)
0  >  org.apache.sysds.test.AutomatedTestBase.runTest(AutomatedTestBase.java:1337)
0  >  org.apache.sysds.test.AutomatedTestBase.runTest(AutomatedTestBase.java:1333)
0  >  org.apache.sysds.test.functions.builtin.BuiltinFFNeuralNetworkTest.testClassificationFit(BuiltinFFNeuralNetworkTest.java:91)
0  >  sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
0  >  sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
0  >  sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
0  >  java.lang.reflect.Method.invoke(Method.java:498)
   >  ... Stopping Stack Trace at JUnit
LEVEL : 1
Exception : class org.apache.sysds.api.DMLException
Message   : failed add function to graph fcall .builtinNS::m_ffTrain
1  >  org.apache.sysds.hops.ipa.FunctionCallGraph.addFunctionOpToGraph(FunctionCallGraph.java:507)
1  >  org.apache.sysds.hops.ipa.FunctionCallGraph.rConstructFunctionCallGraph(FunctionCallGraph.java:451)
1  >  org.apache.sysds.hops.ipa.FunctionCallGraph.rConstructFunctionCallGraph(FunctionCallGraph.java:442)
1  >  org.apache.sysds.hops.ipa.FunctionCallGraph.rConstructFunctionCallGraph(FunctionCallGraph.java:428)
1  >  org.apache.sysds.hops.ipa.FunctionCallGraph.rConstructFunctionCallGraph(FunctionCallGraph.java:399)
1  >  org.apache.sysds.hops.ipa.FunctionCallGraph.addFunctionOpToGraph(FunctionCallGraph.java:487)
1  >  org.apache.sysds.hops.ipa.FunctionCallGraph.rConstructFunctionCallGraph(FunctionCallGraph.java:424)
1  >  org.apache.sysds.hops.ipa.FunctionCallGraph.constructFunctionCallGraph(FunctionCallGraph.java:360)
1  >  org.apache.sysds.hops.ipa.FunctionCallGraph.<init>(FunctionCallGraph.java:94)
1  >  org.apache.sysds.hops.ipa.InterProceduralAnalysis.<init>(InterProceduralAnalysis.java:133)
1  >  org.apache.sysds.parser.DMLTranslator.rewriteHopsDAG(DMLTranslator.java:285)
1  >  org.apache.sysds.api.DMLScript.execute(DMLScript.java:400)
1  >  org.apache.sysds.api.DMLScript.executeScript(DMLScript.java:269)
1  >  org.apache.sysds.test.AutomatedTestBase.main(AutomatedTestBase.java:1464)
1  >  org.apache.sysds.test.AutomatedTestBase.runTest(AutomatedTestBase.java:1420)
1  >  org.apache.sysds.test.AutomatedTestBase.runTest(AutomatedTestBase.java:1341)
1  >  org.apache.sysds.test.AutomatedTestBase.runTest(AutomatedTestBase.java:1337)
1  >  org.apache.sysds.test.AutomatedTestBase.runTest(AutomatedTestBase.java:1333)
1  >  org.apache.sysds.test.functions.builtin.BuiltinFFNeuralNetworkTest.testClassificationFit(BuiltinFFNeuralNetworkTest.java:91)
1  >  sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
1  >  sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
1  >  sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
1  >  java.lang.reflect.Method.invoke(Method.java:498)
   >  ... Stopping Stack Trace at JUnit
LEVEL : 2
Exception : class java.lang.NullPointerException
Message   : null
2  >  org.apache.sysds.hops.ipa.FunctionCallGraph.addFunctionOpToGraph(FunctionCallGraph.java:485)
2  >  org.apache.sysds.hops.ipa.FunctionCallGraph.rConstructFunctionCallGraph(FunctionCallGraph.java:451)
2  >  org.apache.sysds.hops.ipa.FunctionCallGraph.rConstructFunctionCallGraph(FunctionCallGraph.java:442)
2  >  org.apache.sysds.hops.ipa.FunctionCallGraph.rConstructFunctionCallGraph(FunctionCallGraph.java:428)
2  >  org.apache.sysds.hops.ipa.FunctionCallGraph.rConstructFunctionCallGraph(FunctionCallGraph.java:399)
2  >  org.apache.sysds.hops.ipa.FunctionCallGraph.addFunctionOpToGraph(FunctionCallGraph.java:487)
2  >  org.apache.sysds.hops.ipa.FunctionCallGraph.rConstructFunctionCallGraph(FunctionCallGraph.java:424)
2  >  org.apache.sysds.hops.ipa.FunctionCallGraph.constructFunctionCallGraph(FunctionCallGraph.java:360)
2  >  org.apache.sysds.hops.ipa.FunctionCallGraph.<init>(FunctionCallGraph.java:94)
2  >  org.apache.sysds.hops.ipa.InterProceduralAnalysis.<init>(InterProceduralAnalysis.java:133)
2  >  org.apache.sysds.parser.DMLTranslator.rewriteHopsDAG(DMLTranslator.java:285)
2  >  org.apache.sysds.api.DMLScript.execute(DMLScript.java:400)
2  >  org.apache.sysds.api.DMLScript.executeScript(DMLScript.java:269)
2  >  org.apache.sysds.test.AutomatedTestBase.main(AutomatedTestBase.java:1464)
2  >  org.apache.sysds.test.AutomatedTestBase.runTest(AutomatedTestBase.java:1420)
2  >  org.apache.sysds.test.AutomatedTestBase.runTest(AutomatedTestBase.java:1341)
2  >  org.apache.sysds.test.AutomatedTestBase.runTest(AutomatedTestBase.java:1337)
2  >  org.apache.sysds.test.AutomatedTestBase.runTest(AutomatedTestBase.java:1333)
2  >  org.apache.sysds.test.functions.builtin.BuiltinFFNeuralNetworkTest.testClassificationFit(BuiltinFFNeuralNetworkTest.java:91)
2  >  sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2  >  sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2  >  sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2  >  java.lang.reflect.Method.invoke(Method.java:498)
   >  ... Stopping Stack Trace at JUnit

````

For now i would suggest not including parameter server functionality in the builtin, since this provoke the bug and instead focus on the FF, that works.

I have squashed together all your current code in a branch on my fork, if you want to continue on that.

**Comment by aleksapand** [comment]
Thank you for taking a look into it. I will proceed by removing support for paramserv. We will soon finalize LeNet and include it in the PR. Should we continue working on this branch or you prefer working on FFN branch on your fork?

**Comment by Baunsgaard** [comment]
I would suggest to continue work from my branch, but it is up to you, but before you do that then take a look at the differences, i did change a few things that make it a little cleaner, for instance the accuracy calculation in your tests.

**Comment by aleksapand** [comment]
@Baunsgaard we would like to submit the FFN and LeNet tasks for review. You suggested adding the MNIST dataset but the train set in CSV format is a bit over 100MB therefore, we were unable to upload it. We added only a test set and part of it is used to verify the behaviour of LeNet.

**Comment by mboehm7** [comment]
LGTM - thanks for finalizing this PR. During the merge I resolve the merge conflicts (in the future, please rebase and remove unnecessary modifications like cspline), fixed some formatting, and added TODOs for moving the FFN/lenet forward pass from layers into builtin functions.
